### DISCUSSION TITLE
add Prometheus /metrics and k8s /health endpoints

### DISCUSSION THREAD

**Discussion by orefalo** [verdaccio/verdaccio#2079] [Q&A] [answered]
Both can be easily implemented using 
https://github.com/banzaicloud/service-tools
in just a few lines of code

I must point that the following grafana dashboard is preconfigured to visualize the metrics generated by service-tools

https://grafana.com/grafana/dashboards/11159

Last, will open a ticket to adjust the chart at https://github.com/verdaccio/charts

**Comment by juanpicado** [comment]
I suggest create a middleware plugin for it so is used by demand.

https://verdaccio.org/docs/en/plugin-middleware

**Comment by orefalo** [comment]
good point, I know little about plugin-middlewares
Now let me think forward...
Assuming the plugin exists, how would it work with the helm chart?
typically this will controlled via a configuration:
metrics: true

It looks like a plugin needs to be imported manually via an npm command.

Am I missing something?

**Comment by juanpicado** [comment]
> good point, I know little about plugin-middlewares
> Now let me think forward...
> Assuming the plugin exists, how would it work with the helm chart?
> typically this will controlled via a configuration:
> metrics: true
> 
> It looks like a plugin needs to be imported manually via an npm command.
> 
> Am I missing something?

Yes, you might need extend verdaccio image, I'm not sure how k8s fully works. There are examples in our docker examples repo.

**Comment by juanpicado** [comment]
https://github.com/bruceman/verdaccio-hello

**Comment by lilaja** [comment]
@juanpicado  Is there a way to register middleware before built-in endpoints? I mean for example when gathering duration metrics it is important to start timers as soon as possible. 
And judging by  documentation: 
`The middleware are registered after built-in endpoints`
I think it not an option. 
Do you have any recommendation for this use case ?

**Comment by juanpicado** [comment]
Currently not, but next major I was planning to add some sort of lifecycle so some actions can be triggered before and after.

**Comment by lilaja** [comment]
Thanks, I'm looking forward to it!

**Comment by danopia** [comment]
Hello, I've published a Verdaccio plugin that adds basic Prometheus metrics, primarily for HTTP requests. It's available for installation here: https://www.npmjs.com/package/@freighthub/verdaccio-openmetrics and I hope to get the Github repo public relatively soon.

I've found that middlewares are in front of the internal API in the express stack so the API requests (PUTs, etc) do show up. It seems possible to hack the express handler stack from within the plugin to put the metrics first in 4.x however I haven't seen that as super necessary so I haven't implemented that (meaning things like CORS requests are currently not reported).

I'm also looking at what metrics can be emitted about the package database. It seems like 'average tarball size' isn't available, nor 'number of files in avg package' etc so not a ton to go from there. I'm open to suggestions though.

Finally, if a Helm installation wants to configure these metrics, I'm like 90% sure you'd want to have a docker image with the metrics plugin included. Which seems like a pain point unless someday the official image includes a /metrics route.

**Comment by orefalo** [comment]
Thank you @danopia  can't wait to see your package released.

**Comment by danopia** [comment, marked as answer]
My plugin as mentioned above is now open source @ https://github.com/freight-hub/verdaccio-openmetrics

I've relocated the published package to https://www.npmjs.com/package/verdaccio-openmetrics

Hopefully others can find this useful.

---

I am using a Dockerfile like this to create the image:

```Dockerfile
FROM verdaccio/verdaccio:4.10
USER root

ENV NODE_ENV=production
RUN npm install --production \
 && npm install \
        verdaccio-openmetrics@1.0.2 \
 && npm cache clean --force

USER verdaccio
```

.. and then giving that image tag to Helm to be deployed.

**Comment by orefalo** [comment]
thank you so much - will definitely implement over the weekend.

**Reply by micaelmbagira** [reply]
@danopia it looks like the plugin is not compatible with verdaccio 5. Do you know what would need to be changed?

**Reply by danopia** [reply]
> @danopia it looks like the plugin is not compatible with verdaccio 5.

Thanks for the heads up. I haven't taken the 5.x upgrade yet on my installation. I'm on vacation for the next two weeks, so if there's a reasonable change to address this, please file a PR for my team to triage/release - otherwise I'll look at it when I'm back at a desk :)

**Reply by danopia** [reply]
> @danopia it looks like the plugin is not compatible with verdaccio 5. Do you know what would need to be changed?

Hi, just following up on this - I've had a 5.x-targetting release [v1.2.0](https://github.com/freight-hub/verdaccio-openmetrics/releases/tag/v1.2.0) up since December, but I also saw that it didn't really seem to matter when updating my own verdaccio to 5.x. I hope you got your upgrade sorted out too. :)

**Comment by edclement** [comment]
@danopia - cool plugin! 

For others involved in this discussion who are looking for Prometheus compatible metrics that are more focused on incoming requests and package downloads, [XLTS.dev](https://xlts.dev) maintains the [verdaccio-prometheus-middleware](https://github.com/xlts-dev/verdaccio-prometheus-middleware) plugin. It is published to [npm](https://www.npmjs.com/package/@xlts.dev/verdaccio-prometheus-middleware) as well. We hope this can be useful to other Verdaccio users as well.

**Reply by juanpicado** [reply]
You might want to add this to the website in the Kubernetes or Docker one, almost no traffic here.

**Reply by edclement** [reply]
👍🏼 Yeah, it's been on my TODO list to open a PR to do so.
